File: Sean的leetcode/FB/Subsets II.py
###
# Time Complexity: O(2^n)
# Space Complexity: O(n)
###
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums:
            return []
        re = []
        nums.sort()
        self.dfs(nums, 0, [], re)
        return re
        
    def dfs(self, nums, idx, one_sol, re):
        if one_sol not in re:
            re.append(one_sol)
        for i in xrange(idx, len(nums)):
            if i == idx or nums[i] != nums[i-1]:
                self.dfs(nums, i+1, one_sol + [nums[i]], re)
        

File: MyLeetCode/Leetcode/Subsets II.cpp
class Solution {
public:
    void genSub(vector<int>& nums, int start, vector<int> &one_sol, vector<vector<int> > &res)
    {
        for (int i = start; i < nums.size(); i++)
        {
            one_sol.push_back(nums[i]);
            res.push_back(one_sol);
            genSub(nums, i+1, one_sol, res);
            one_sol.pop_back();

            while (i + 1 < nums.size() && nums[i] == nums[i+1])
                i++;
        }
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> one_sol;
        vector<vector<int> > res;

        res.push_back(one_sol);
        sort(nums.begin(), nums.end());

        genSub(nums, 0, one_sol, res);
        return res;
    }
};